src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -411,28 +413,43 @@ public void ProduceContinue()
             }
         }
 
-        public void ProduceStart(bool immediate = true, bool appCompleted = false)
+        public async Task ProduceStartAndFireOnStarting(bool immediate = true)
+        {
+            if (_responseStarted) return;
+
+            await FireOnStarting();
+           
+            if (_applicationFailed) 
+            {
+                throw new ObjectDisposedException(typeof(Frame).FullName);
```

Comment:
We should at least include a message saying that the response was aborted due to an application exception. We could even include the exception as a nested exception.